ext/opcache/jit/zend_jit_arm64.dasc

```diff
@@ -483,10 +483,100 @@ static int logical_immediate_p (uint64_t value, uint32_t reg_size)
 ||	}
 |.endmacro
 
+#ifdef __APPLE__
+|.macro SAVE_REGS
+|	sub sp, sp, #(16*21)
+|
+|	stp d30, d31, [sp]
+|	stp d28, d29, [sp, #16]
+|	stp d26, d27, [sp, #(16*2)]
+|	stp d24, d25, [sp, #(16*3)]
+|	stp d22, d23, [sp, #(16*4)]
+|	stp d20, d21, [sp, #(16*5)]
+|	stp d18, d19, [sp, #(16*6)]
+|	stp d16, d17, [sp, #(16*7)]
+|	// Callee-saved registers, d8 to d15, can be omitted.
+|	stp d6, d7, [sp, #(16*8)]
+|	stp d4, d5, [sp, #(16*9)]
+|	stp d2, d3, [sp, #(16*10)]
+|	stp d0, d1, [sp, #(16*11)]
+|
+|	stp x29, x30, [sp, #(16*12)]
+|	// Callee-saved registers, x19 to x28, can be omitted.
+|	// Temporary registers TMP1 to TMP3, i.e. x15 to 17, can be omitted.
+|	stp x14, x18, [sp, #(16*13)]
+|	stp x12, x13, [sp, #(16*14)]
+|	stp x10, x11, [sp, #(16*15)]
+|	stp x8, x9, [sp, #(16*16)]
+|	stp x6, x7, [sp, #(16*17)]
+|	stp x4, x5, [sp, #(16*18)]
+|	stp x2, x3, [sp, #(16*19)]
+|	stp x0, x1, [sp, #(16*20)]
+|.endmacro
+
+|.macro RESTORE_REGS
+|	ldp d30, d31, [sp]
+|	ldp d28, d29, [sp, #16]
+|	ldp d26, d27, [sp, #(16*2)]
+|	ldp d24, d25, [sp, #(16*3)]
+|	ldp d22, d23, [sp, #(16*4)]
+|	ldp d20, d21, [sp, #(16*5)]
+|	ldp d18, d19, [sp, #(16*6)]
+|	ldp d16, d17, [sp, #(16*7)]
+|	// Callee-saved registers, d8 to d15, can be omitted.
+|	ldp d6, d7, [sp, #(16*8)]
+|	ldp d4, d5, [sp, #(16*9)]
+|	ldp d2, d3, [sp, #(16*10)]
+|	ldp d0, d1, [sp, #(16*11)]
+|
+|	ldp x29, x30, [sp, #(16*12)]
+|	// Callee-saved registers, x19 to x28, can be omitted.
+|	// Temporary registers TMP1 to TMP3, i.e. x15 to 17, can be omitted.
+|	ldp x14, x18, [sp, #(16*13)]
+|	ldp x12, x13, [sp, #(16*14)]
+|	ldp x10, x11, [sp, #(16*15)]
+|	ldp x8, x9, [sp, #(16*16)]
+|	ldp x6, x7, [sp, #(16*17)]
+|	ldp x4, x5, [sp, #(16*18)]
+|	ldp x2, x3, [sp, #(16*19)]
+|	ldp x0, x1, [sp, #(16*20)]
+|
+|	add sp, sp, #(16*21)
+|.endmacro
+#endif
+
 |.macro LOAD_TSRM_CACHE, reg
+||#ifdef __APPLE__
+|	// TODO: Get a better name for "tsrm_ls_cache_tcb_offset"?
+|	// TODO: In Linux only TMP3 is used particularly for ZTS, whereas TMP1 and TMP2 are also used in macOS.
+|	LOAD_ADDR TMP3, tsrm_ls_cache_tcb_offset
+|	.long 0xd53bd06f // TODO: hard-coded: mrs TMP1, tpidrro_el0
+|	and TMP1, TMP1, #0xfffffffffffffff8
+|	ldr TMP2, [TMP3, #0x8]
+|	ldr TMP1, [TMP1, TMP2, lsl #3]
+|	cbz TMP1, >4     // TODO: There is conflict for local labels. A number of test cases would fail.
```

Comment:
Yes. Agree. The slow path should be avoided. But may I ask where is the proper place to access `tsrm_ls_cache`?

`0xfffffffffffffff8` is used as CPU number mask. See https://opensource.apple.com/source/xnu/xnu-4570.1.46/osfmk/arm64/cswitch.s.auto.html, the "set_thread_registers" section.

`ldr TMP2, [TMP3, #0x8]` is to get the "key" and `ldr TMP2, [TMP3, #0x10]` is used to get the "offset". See `struct TLVDescriptor` from https://opensource.apple.com/source/dyld/dyld-195.6/src/threadLocalVariables.c.auto.html

I'm afraid I didn't get `lsl #3` either.

Yes. I agree. "key" and "offset" are both constant and we can get them in the TSRM.c code.

Reply:
Actually, we may keep TSRM.c and do something like this.
```
mrs TMP1, tpidrro_el0
SAFE_MEM_ACC_WITH_UOFFSET ldr, TMP1, TMP1, ((TLVDescriptor*)(tsrm_ls_cache_tcb_offset))->key << 3, TMP2
SAFE_MEM_ACC_WITH_UOFFSET ldr, TMP1, TMP1, ((TLVDescriptor*)(tsrm_ls_cache_tcb_offset))->offset, TMP2
```

may be need also add ``and TMP1, TMP1, #0xfffffffffffffff8``

Reply:
I got your point. I will update the code soon.

Another question for me is where is the best place for each thread to access  thread local varaible`_tsrm_ls_cache` ?

I guess `zend_jit_setup()` -> `tsrm_get_ls_cache_tcb_offset()` is executed only once for each process, not for each thread.

Reply:
``_tsrm_ls_cache`` is guaranteed to be accessed before any VM or JIT code, during request startup, e.g. in init_executor()

zend_jit_setup() executed once, but in case TLVDescriptor is immutable, this should be fine.

Reply:
`TLVDescriptor` for variable `_tsrm_ls_cache` is immutable, but the value in `TMP1` at line 556 varies for different threads. This value is also `SAFE_MEM_ACC_WITH_UOFFSET ldr, TMP1, TMP1, ((TLVDescriptor*)(tsrm_ls_cache_tcb_offset))->key << 3, TMP2` for your update.

If `TMP1` is zero at line 556, slow path would be invoked.

Here I think we should guarantee this `TMP1` at line 556 (the value stored in 
`TMP1 + ((TLVDescriptor*)tsrm_ls_cache_tcb_offset))->key << 3)` ) for **each** thread is initialized in advance. Right?

Reply:
> `TLVDescriptor` for variable `_tsrm_ls_cache` is immutable, but the value in `TMP1` at line 556 varies for different threads. This value is also `SAFE_MEM_ACC_WITH_UOFFSET ldr, TMP1, TMP1, ((TLVDescriptor*)(tsrm_ls_cache_tcb_offset))->key << 3, TMP2` for your update.

This is expected, but I don't see any problems.

> If `TMP1` is zero at line 556, slow path would be invoked.
> 
> Here I think we should guarantee this `TMP1` at line 556 (the value stored in
> `TMP1 + ((TLVDescriptor*)tsrm_ls_cache_tcb_offset))->key << 3)` ) for **each** thread is initialized in advance. Right?

In my opinion, it can't be zero at the moment when JIT-ed code is executed.